### PR TITLE
GNN Model with 93% Accuracy for Facebook Page-Page Network Node Classification with TSNE Visualization

### PR DESCRIPTION
**Pull request by liammulhern**
# GNN Model with 93% Accuracy for Facebook Page-Page Network Node Classification with TSNE Visualization
This project introduces a multi-layer graph neural network (GNN) for semi-supervised, multi-class node classification on the Facebook Large Page-Page Network dataset, achieving **93.14% accuracy**. The network classifies nodes (representing Facebook pages) into four categories: **Politicians, Government Organizations, Television Shows, and Companies**.

# Key features of PR:

## Modules:
- `dataset.py`: Loads and preprocesses data.
- `main.py`: CLI for training and inference.
- `modules.py`: Defines GNN architecture.
- `train.py`: Manages training, validation, and metric logging.
- `predict.py`: Runs model inference and visualizations.

## Execution:
Supports training (`--train` `--save` `--load`), inference (`--inference <index>`), and visualization (`--display`) through CLI.

## GNN Architecture:
1. Uses multilayer perceptrons (MLPs) and sparse layers, transforming node features into learned embeddings for classification.
2. Incorporates ReLU activation and log softmax for output.

## Training:
- Learning Rate: 1e-4
- Epochs: 100
- Optimizer: Adam
- Loss: Cross-Entropy

## Results:
Achieves **93.14% accuracy**; training and validation metrics show potential overfitting.
TSNE visualizations show clearer clustering post-training, indicating successful categorization.

### CODE REVIEW

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Normal

Readme Overall: Good
- Project Overview: Good
- Model Documented/Explained: Fair. Mentions the use of the sparse layer, though the implemented model does not use this layer ( torch_geometric.nn import GCNConv).  
- Data + Pre-processing: Not done. Used 128 embedding as model input. This is ok.
- Training / Loss Curve(s): OK, not specified train or test accuracy.  Shows loss curves.  Marginal overfit, as gap seems increasing.
- Result Demonstration: Good
- References: Good
- Setup/Reproducibility/Requirements: Good

Functionality:
- Multi-layer GCN: Good, 2 layers. Fair. Uses external library (torch_geometric.nn GCNConv). 
- Semi supervised: Good
- Accuracy: OK. Not mentioned if its Test or Train accuracy.
- TSNE or UMAP: Good.
- Discussion: Fair.

Code:
- Consistent with Results: OK.
- File Structure: Good. Redundant code in repo - SparseLayer
- Commenting: Good
- Commit frequency: Good
- Commit messages: Good
- Reusable Modular code/Design: Good

Other Comments:
No early stop.
No hyper parameter search.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Adequate design and implementation. Redundant code in repo - SparseLayer. </td><td>-1</td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Some overfitting seen. No early stop</td><td>-2</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Normal. GNN Task.</td><td>-5</td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. Commits contain previous GAN work, and only 2 days of GCN. This is fine.</td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Acceptable. </td><td>-1</td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF NOT submitted. </td><td>-2</td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). </td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-11</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).
